HTTP1: Read data on backend connection reset

Algunenano opened the pull request:
If the backend closes a connection, try to read any available data to see if there is information about it.
This is to deal with common behaviour on some web servers that close the connection as soon as they find an issue with it, for example a large content length.

References https://github.com/varnishcache/varnish-cache/issues/2332

The PR is missing tests (I need to learn how they work and how to recreate this setup within the tests) but I wanted to open it in order to receive feedback as soon as possible as this is my first contribution to the project.

An example with a server that returns `400 Bad Request` based on Content-Length and closes the connection to interrupt the upload:

Before:

```
*              << BeReq    >>   3         
-            3 Begin          b bereq 2 pass
-            3 VCL_use        b boot
-            3 Timestamp      b Start: 1630953265.053809 0.000000 0.000000
-            3 BereqMethod    b POST
-            3 BereqURL       b /v0/datasources?name=stock_prices_1M&mode=append
-            3 BereqProtocol  b HTTP/1.1
-            3 BereqHeader    b Host: localhost:6081
-            3 BereqHeader    b User-Agent: curl/7.78.0
-            3 BereqHeader    b Accept: */*
-            3 BereqHeader    b Authorization: Bearer p.eyJ1IjogImRlMzExNzI4LWY3M2ItNGRmOS05ZWFlLTA2Yzc0Y2VkZGFlNCIsICJpZCI6ICIzYTFiYTQzMC0zYzNlLTRlN2QtYjcyZS0wOTljMjE1NmNlYTQifQ.1kpgmxugmlRk6WYRB38Z6bQskT9d9qnwpEb0BAz01gM
-            3 BereqHeader    b Content-Length: 84248385
-            3 BereqHeader    b Content-Type: application/x-www-form-urlencoded
-            3 BereqHeader    b X-Forwarded-For: 127.0.0.1
-            3 BereqHeader    b X-Varnish: 3
-            3 VCL_call       b BACKEND_FETCH
-            3 VCL_return     b fetch
-            3 Timestamp      b Fetch: 1630953265.053821 0.000012 0.000012
-            3 Timestamp      b Connected: 1630953265.053908 0.000098 0.000086
-            3 BackendOpen    b 26 server1 127.0.0.1 8001 127.0.0.1 56358 connect
-            3 FetchError     b req.body read error: 104 (Connection reset by peer)
-            3 FetchError     b backend write error: 104 (Connection reset by peer)
-            3 Timestamp      b Bereq: 1630953265.056756 0.002947 0.002848
-            3 BackendClose   b 26 server1 close
-            3 Timestamp      b Beresp: 1630953265.056763 0.002954 0.000006
-            3 Timestamp      b Error: 1630953265.056764 0.002954 0.000000
-            3 BerespProtocol b HTTP/1.1
-            3 BerespStatus   b 503
-            3 BerespReason   b Backend fetch failed
-            3 BerespHeader   b Date: Mon, 06 Sep 2021 18:34:25 GMT
-            3 BerespHeader   b Server: Varnish
-            3 VCL_call       b BACKEND_ERROR
-            3 BerespHeader   b Content-Type: text/html; charset=utf-8
-            3 BerespHeader   b Retry-After: 5
-            3 VCL_return     b deliver
-            3 Storage        b malloc Transient
-            3 Length         b 278
-            3 BereqAcct      b 435 2652075 2652510 0 0 0
-            3 End            b 

*              << Request  >>   2         
-            2 Begin          c req 1 rxreq
-            2 Timestamp      c Start: 1630953265.053717 0.000000 0.000000
-            2 Timestamp      c Req: 1630953265.053717 0.000000 0.000000
-            2 VCL_use        c boot
-            2 ReqUnset       c Expect: 100-continue
-            2 ReqStart       c 127.0.0.1 56556 a0
-            2 ReqMethod      c POST
-            2 ReqURL         c /v0/datasources?name=stock_prices_1M&mode=append
-            2 ReqProtocol    c HTTP/1.1
-            2 ReqHeader      c Host: localhost:6081
-            2 ReqHeader      c User-Agent: curl/7.78.0
-            2 ReqHeader      c Accept: */*
-            2 ReqHeader      c Authorization: Bearer p.eyJ1IjogImRlMzExNzI4LWY3M2ItNGRmOS05ZWFlLTA2Yzc0Y2VkZGFlNCIsICJpZCI6ICIzYTFiYTQzMC0zYzNlLTRlN2QtYjcyZS0wOTljMjE1NmNlYTQifQ.1kpgmxugmlRk6WYRB38Z6bQskT9d9qnwpEb0BAz01gM
-            2 ReqHeader      c Content-Length: 84248385
-            2 ReqHeader      c Content-Type: application/x-www-form-urlencoded
-            2 ReqHeader      c X-Forwarded-For: 127.0.0.1
-            2 VCL_call       c RECV
-            2 VCL_return     c pass
-            2 RespProtocol   c HTTP/1.1
-            2 RespStatus     c 100
-            2 RespReason     c Continue
-            2 VCL_call       c HASH
-            2 VCL_return     c lookup
-            2 VCL_call       c PASS
-            2 VCL_return     c fetch
-            2 Link           c bereq 3 pass
-            2 Storage        c malloc Transient
-            2 Timestamp      c ReqBody: 1630953265.056595 0.002877 0.002877
-            2 Timestamp      c Fetch: 1630953265.056799 0.003082 0.000204
-            2 RespProtocol   c HTTP/1.1
-            2 RespStatus     c 503
-            2 RespReason     c Backend fetch failed
-            2 RespHeader     c Date: Mon, 06 Sep 2021 18:34:25 GMT
-            2 RespHeader     c Server: Varnish
-            2 RespHeader     c Content-Type: text/html; charset=utf-8
-            2 RespHeader     c Retry-After: 5
-            2 RespHeader     c X-Varnish: 2
-            2 RespHeader     c Age: 0
-            2 RespHeader     c Via: 1.1 varnish (Varnish/6.6)
-            2 VCL_call       c DELIVER
-            2 RespUnset      c Via: 1.1 varnish (Varnish/6.6)
-            2 RespUnset      c X-Varnish: 2
-            2 RespUnset      c Age: 0
-            2 VCL_return     c deliver
-            2 Timestamp      c Process: 1630953265.056806 0.003089 0.000007
-            2 Filters        c 
-            2 RespHeader     c Content-Length: 278
-            2 RespHeader     c Connection: close
-            2 Timestamp      c Resp: 1630953265.056848 0.003131 0.000041
-            2 ReqAcct        c 415 4390912 4391327 212 278 490
-            2 End            c 

*              << Session  >>   1         
-            1 Begin          c sess 0 HTTP/1
-            1 SessOpen       c 127.0.0.1 56556 a0 127.0.0.1 6081 1630953265.053680 24
-            1 Link           c req 2 rxreq
-            1 SessClose      c RX_BODY 0.003
-            1 End            c 

```


After:

```
*              << BeReq    >>   3         
-            3 Begin          b bereq 2 pass
-            3 VCL_use        b boot
-            3 Timestamp      b Start: 1630953155.802987 0.000000 0.000000
-            3 BereqMethod    b POST
-            3 BereqURL       b /v0/datasources?name=stock_prices_1M&mode=append
-            3 BereqProtocol  b HTTP/1.1
-            3 BereqHeader    b Host: localhost:6081
-            3 BereqHeader    b User-Agent: curl/7.78.0
-            3 BereqHeader    b Accept: */*
-            3 BereqHeader    b Authorization: Bearer p.eyJ1IjogImRlMzExNzI4LWY3M2ItNGRmOS05ZWFlLTA2Yzc0Y2VkZGFlNCIsICJpZCI6ICIzYTFiYTQzMC0zYzNlLTRlN2QtYjcyZS0wOTljMjE1NmNlYTQifQ.1kpgmxugmlRk6WYRB38Z6bQskT9d9qnwpEb0BAz01gM
-            3 BereqHeader    b Content-Length: 84248385
-            3 BereqHeader    b Content-Type: application/x-www-form-urlencoded
-            3 BereqHeader    b X-Forwarded-For: 127.0.0.1
-            3 BereqHeader    b X-Varnish: 3
-            3 VCL_call       b BACKEND_FETCH
-            3 VCL_return     b fetch
-            3 Timestamp      b Fetch: 1630953155.803004 0.000017 0.000017
-            3 Timestamp      b Connected: 1630953155.803081 0.000093 0.000076
-            3 BackendOpen    b 26 server1 127.0.0.1 8001 127.0.0.1 56304 connect
-            3 FetchError     b req.body read error: 104 (Connection reset by peer)
-            3 Timestamp      b Bereq: 1630953155.805908 0.002920 0.002826
-            3 FetchError     b backend write error: 104 (Connection reset by peer) but status received: 400
-            3 Timestamp      b Beresp: 1630953155.805916 0.002928 0.000008
-            3 BerespProtocol b HTTP/1.1
-            3 BerespStatus   b 400
-            3 BerespReason   b Bad Request
-            3 BerespHeader   b Date: Mon, 06 Sep 2021 18:32:35 GMT
-            3 VCL_call       b BACKEND_RESPONSE
-            3 TTL            b VCL 60 0 0 1630953156 uncacheable
-            3 TTL            b VCL 60 60 0 1630953156 uncacheable
-            3 VCL_return     b deliver
-            3 Timestamp      b Process: 1630953155.805925 0.002938 0.000009
-            3 Filters        b 
-            3 Storage        b malloc Transient
-            3 Fetch_Body     b 4 eof stream
-            3 BackendClose   b 26 server1 close
-            3 Timestamp      b BerespBody: 1630953155.805949 0.002961 0.000023
-            3 Length         b 0
-            3 BereqAcct      b 435 2652075 2652510 28 0 28
-            3 End            b 

*              << Request  >>   2         
-            2 Begin          c req 1 rxreq
-            2 Timestamp      c Start: 1630953155.802913 0.000000 0.000000
-            2 Timestamp      c Req: 1630953155.802913 0.000000 0.000000
-            2 VCL_use        c boot
-            2 ReqUnset       c Expect: 100-continue
-            2 ReqStart       c 127.0.0.1 56554 a0
-            2 ReqMethod      c POST
-            2 ReqURL         c /v0/datasources?name=stock_prices_1M&mode=append
-            2 ReqProtocol    c HTTP/1.1
-            2 ReqHeader      c Host: localhost:6081
-            2 ReqHeader      c User-Agent: curl/7.78.0
-            2 ReqHeader      c Accept: */*
-            2 ReqHeader      c Authorization: Bearer p.eyJ1IjogImRlMzExNzI4LWY3M2ItNGRmOS05ZWFlLTA2Yzc0Y2VkZGFlNCIsICJpZCI6ICIzYTFiYTQzMC0zYzNlLTRlN2QtYjcyZS0wOTljMjE1NmNlYTQifQ.1kpgmxugmlRk6WYRB38Z6bQskT9d9qnwpEb0BAz01gM
-            2 ReqHeader      c Content-Length: 84248385
-            2 ReqHeader      c Content-Type: application/x-www-form-urlencoded
-            2 ReqHeader      c X-Forwarded-For: 127.0.0.1
-            2 VCL_call       c RECV
-            2 VCL_return     c pass
-            2 RespProtocol   c HTTP/1.1
-            2 RespStatus     c 100
-            2 RespReason     c Continue
-            2 VCL_call       c HASH
-            2 VCL_return     c lookup
-            2 VCL_call       c PASS
-            2 VCL_return     c fetch
-            2 Link           c bereq 3 pass
-            2 Storage        c malloc Transient
-            2 Timestamp      c ReqBody: 1630953155.805755 0.002841 0.002841
-            2 Timestamp      c Fetch: 1630953155.805938 0.003024 0.000182
-            2 RespProtocol   c HTTP/1.1
-            2 RespStatus     c 400
-            2 RespReason     c Bad Request
-            2 RespHeader     c Date: Mon, 06 Sep 2021 18:32:35 GMT
-            2 RespHeader     c X-Varnish: 2
-            2 RespHeader     c Age: 0
-            2 RespHeader     c Via: 1.1 varnish (Varnish/6.6)
-            2 VCL_call       c DELIVER
-            2 RespUnset      c Via: 1.1 varnish (Varnish/6.6)
-            2 RespUnset      c X-Varnish: 2
-            2 RespUnset      c Age: 0
-            2 VCL_return     c deliver
-            2 Timestamp      c Process: 1630953155.805943 0.003029 0.000005
-            2 Filters        c 
-            2 RespHeader     c Connection: close
-            2 RespHeader     c Transfer-Encoding: chunked
-            2 Timestamp      c Resp: 1630953155.805983 0.003070 0.000040
-            2 ReqAcct        c 415 4193191 4193606 137 0 137
-            2 End            c 

*              << Session  >>   1         
-            1 Begin          c sess 0 HTTP/1
-            1 SessOpen       c 127.0.0.1 56554 a0 127.0.0.1 6081 1630953155.802880 22
-            1 Link           c req 2 rxreq
-            1 SessClose      c RX_BODY 0.003
-            1 End            c 
```

Dridi commented:
Second try:

```
varnishtest 3686

barrier b1 cond 2

server s1 {
        rxreqhdrs
        txresp -status 400 -body "Invalid request"
} -start

varnish v1 -vcl+backend { } -start

client c1 {
        txreq -method POST -hdr "Transfer-Encoding: chunked"
        chunkedlen 100
        barrier b1 sync
        chunkedlen 100
        chunkedlen 0
        rxresp
        expect resp.status == 400
        expect resp.body == "Invalid request"
} -start

server s1 -wait
barrier b1 sync
client c1 -wait
```

My first attempt with a content length failed.

Dridi commented:
```
varnishtest 3686

barrier b1 cond 2

server s1 {
        rxreqhdrs
        txresp -status 400 -body "Invalid request"
} -start

varnish v1 -vcl+backend { } -start

client c1 {
        txreq -method POST -hdr "Content-Length: 2000"
        send_n 100 0123456789
        barrier b1 sync
        send_n 100 0123456789
        rxresp
        expect resp.status == 400
        expect resp.body == "Invalid request"
} -start

server s1 -wait
barrier b1 sync
client c1 -wait
```

Larger request body did the trick, `flush_head` would have too.

nigoroll commented:
:+1: on the general idea, have not looked close enough at the patch yet to comment in more detail

Dridi commented:
So, this poses a conceptual problem: since we failed to send the `bereq`, do we transition to `vcl_backend_error` with a non-synthetic `beresp` or do we go to `vcl_backend_response` instead?

I think we should:

- go to `vcl_backend_response`
- add `bereq.send_failed` and `beresp.fetch_failed` flags or something similar
  - available in `vcl_backend_error` too
- have the built-in `vcl_backend_response` raise `beresp.uncacheable` if we failed to send the `bereq`
  - although, with a request body there's a fair chance that `bereq` was already uncacheable

Algunenano commented:
> I think we should:

Everything you mention makes sense to me. If the server sent a valid response before closing the connection, go to `vcl_backend_response` with a flag(s) to mark that; if it didn't, then go to the normal path `vcl_backend_error`. Both cases should set `beresp.uncacheable` too. What I don't know yet is how to do all that :D 

Since we are considering adding new flags this raises an extra question. Should this cover HTTP2 too?
The original comes from the fact that under HTTP/1.1 ([rft2616](https://datatracker.ietf.org/doc/html/rfc2616) or [rfc7230](https://datatracker.ietf.org/doc/html/rfc7230)) the server may close the connection to to prevent the client from continuing the request for one of multiple reasons like too large payload, an unknown encoding, etc.
On the other hand, HTTP/2 ([rfc7540](https://datatracker.ietf.org/doc/html/rfc7540#section-5.4)) does have an error handling mechanism so the server can terminate a stream without having the close the connection.

Dridi commented:
Your observations are basically what I mean by "or something similar". We don't have h2 backends, but one may write their own backend implementation in a VMOD. So we'll need to think about what to expose in VCL and figure out the semantics.

Your patch focuses (mostly) on the HTTP/1 fetch so we can still review that part independently.

Algunenano commented:
I've added a proposal adding `beresp.send_failed` `beresp.fetch_failed` available at `vcl_backend_response` and `vcl_backend_error` with some tests (I wasn't sure how to implement multiple of them in the same file without overcomplicating things). We can discuss further about the semantics and so on, but that's something that works and could be easily modified.

nigoroll commented:
👍🏽 except for the `bereq.send_failed` nit

bsdphk commented:
We need to ensure that the fd either A) does not make it back to a waiter or B) that the waiter copes.

Doing A) by closing is probably the safest.

Algunenano commented:
* Changed `beresp.send_failed` to  `bereq.send_failed`
* Further modified `vcl_builtin_backend_response` to add a `Connection: close` header.
* Added a check to confirm that the backend connection is closed; expects `BackendClose    b 26 s1 close` and not `BackendClose    b 26 s1 recycle`.

Algunenano commented:
I see a few tests failing locally (`tests/f00001.log tests/m00000.log tests/r01356.log tests/r01562.log tests/r03556.log`). I'll review them tomorrow.

Dridi commented:
> Further modified `vcl_builtin_backend_response` to add a `Connection: close` header.

I think what we want is rather that we don't attempt to recycle the connection even without VCL intervention.

Algunenano commented:
> I think what we want is rather that we don't attempt to recycle the connection even without VCL intervention.

Fair enough. It should be covered buy the other tests so I'll remove it.

Algunenano commented:
About the failing tests:
* `tests/m00000.log` fails for me also in master/HEAD (`Not true: vmods (3) == 0 (0)`). Maybe related to flags or the local machine. I'm not too worried about it.
* The rest of the tests fail because I moved the send_failed check when adding the new vcl getters and it was treating `if (sc == SC_NULL && i < 0)` also as a send error, which as far as I can see, it's not a desirable behaviour.

I've minimized the changes against the current master/HEAD and pushed the fixes.

Dridi commented:
Yesterday felt bad about introducing conflicts in this pull request in #3753 and 948cf26deb09848aade89d7f0c1f7e7a669f592a so I decided to review it and rebase the branch myself to spare you the conflict resolution. I found a few problems and loose ends with this branch and decided to try it from a different angle in #3761.

Please give it a try if you get a chance.

Algunenano commented:
Superseeded by #3761
